### PR TITLE
DBP: adding other actions

### PR DESCRIPTION
**Pull request by shakyShane**
https://app.asana.com/0/0/1206000767975136/f

**Feature Summary**

Native platforms will send one of the following instructions, one at a time:

- `navigate`
- `extract`
- `click` <- already added in a previous PR
- `expectation`
- `fillForm`
- `getCaptchaInfo`
- `solveCaptcha`

Each of them has a corresponding function that will be executed, and will return either a SuccessResponse or ErrorResponse to native.

The flow looks like this: (taken from the markdown file in the PR)

image

### CODE REVIEW

**Comment by shakyShane** [comment]
Current dependencies on/for this PR:
* main
  * **PR #843** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/843?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #844** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/844?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #845** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/845?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/844?utm_source=stack-comment).

**Comment by shakyShane** [comment]
@muodov this adds all the other actions, the next PR has all the tests.

Please let me know if you'd like me to break this down further, I'd be happy to do so if it's easier to review

File: src/features/broker-protection/buildUrl.js
Comment (brianhall) [inline src/features/broker-protection/buildUrl.js:35]: Is this just to catch an incorrectly formatted URL? If so - nice! 😄

File: src/features/broker-protection/comparison-functions.js
Comment (brianhall) [inline src/features/broker-protection/comparison-functions.js:299]: Do we want this and the state abbreviations in this file? Seems like something we could kick to a constants directory so it's easier to find?

File: src/features/broker-protection/comparison-functions.js
Comment (brianhall) [inline src/features/broker-protection/comparison-functions.js:8]: @shakyShane I don't feel really strongly about this, but I would almost prefer each of these functions to be separate files in a utils folder, just from a code organization standpoint. They're pretty involved, so it would help to split up this monolith a bit.

File: src/features/broker-protection/fillForm.js
Comment (brianhall) [inline src/features/broker-protection/fillForm.js:19]: nit: Do we need this in here?

**Review by brianhall — DISMISSED** [review]
This LGTM, no blockers, but a few questions/comments.

File: src/features/broker-protection/comparison-functions.js
Comment (shakyShane) [reply src/features/broker-protection/comparison-functions.js:299]: agreed, nice idea

File: unit-test/broker-protection.js
Comment (shakyShane) [inline unit-test/broker-protection.js:373]: these property-based tests were added to help me stress-test the URL replacement stuff (interpolating data). It helped me catch a couple of bugs, and in general it would be nice to have fast-check around in C-S-S to use for other fuzzing-style tests

File: src/features/broker-protection/comparison-functions.js
Comment (shakyShane) [reply src/features/broker-protection/comparison-functions.js]: Done!

File: src/features/broker-protection/buildUrl.js
Comment (shakyShane) [reply src/features/broker-protection/buildUrl.js]: yep!

File: unit-test/broker-protection.js
Comment (jonathanKingston) [reply unit-test/broker-protection.js:373]: Are we using fast-check anywhere else in JS land? If not I think it's an open a new tech thing as it's sizable dependency.

File: package.json
Comment (jonathanKingston) [inline package.json:80]: This looks pretty unmaintained but given the size probably acceptable risk.

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (jonathanKingston) [inline src/features/broker-protection/actions/build-url-transforms.js:74]: What kind of processing?

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (jonathanKingston) [inline src/features/broker-protection/actions/build-url-transforms.js:121]: Is this for processing 'one' segment of the URL?

```suggestion
export function processOneURLSegment (input, action, userData) {
```

File: src/features/broker-protection/actions/extract.js
Comment (muodov) [inline src/features/broker-protection/actions/extract.js:79]: hm, `evalueatedValue` may be an array (if it's produced by `findFromElements()`), but `extractValue` seems to expect a single string (at least in some extractors such as `name`). Am I missing something?

File: src/features/broker-protection/actions/extract.js
Comment (muodov) [inline src/features/broker-protection/actions/extract.js:139]: notice: I just glanced over this and underlying functions as they are largely heuristics and data massage, so I can't really judge the correctness

File: unit-test/broker-protection.js
Comment (shakyShane) [reply unit-test/broker-protection.js:373]: AFAIK it's the most widely used in the JS eco-system (likely not a huge space, given that many don't reach for this kind of testing)

We have it in autofill (password generation) - but I believe that was agreed during the code-review, rather than a new-tech discussion.

We could remove these tests (I applied the fixes they raised) and then come back after we've done the official channels :)

File: package.json
Comment (shakyShane) [reply package.json:80]: yeah agreed - I spoke about this at the time. The idea was that it would come up in code review and we'd decide to either keep a dependency, or vendor it in etc.

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url-transforms.js:74]: oops! typo, will fix. should just be 'produce a new one'

File: src/features/broker-protection/actions/fill-form.js
Comment (muodov) [inline src/features/broker-protection/actions/fill-form.js:45]: we could also capture this setter in captured-globals, so it's even safer

File: src/features/broker-protection/actions/fill-form.js
Comment (muodov) [inline src/features/broker-protection/actions/fill-form.js:5]: Is this a copypaste glitch?

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url-transforms.js:17]: No - these are never in that format.

File: src/features/broker-protection/actions/build-url.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url.js:35]: Nope - these are never in that format.

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url-transforms.js:123]: nice, yeah I'll test both of those suggestions!

File: src/features/broker-protection/actions/extract.js
Comment (shakyShane) [reply src/features/broker-protection/actions/extract.js:79]: No, I think this just needs be a bit more explicit, let me re-work it

File: package.json
Comment (jonathanKingston) [reply package.json:80]: Maybe it shouldn't be a dev dep though as we're integrating it into the generated code right?

File: unit-test/broker-protection.js
Comment (jonathanKingston) [reply unit-test/broker-protection.js:373]: > AFAIK it's the most widely used in the JS eco-system 

I meant our ecosystem.

> We could remove these tests (I applied the fixes they raised) and then come back after we've done the official channels :)

I think we're good if used elsewhere and it's only for testing.

File: package.json
Comment (shakyShane) [reply package.json:80]: Ahh yeah that's a good point.

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url-transforms.js]: I renamed the function completely to make this clearer :)

File: src/features/broker-protection/actions/fill-form.js
Comment (shakyShane) [reply src/features/broker-protection/actions/fill-form.js:5]: yep, removing :)

File: src/features/broker-protection/actions/fill-form.js
Comment (shakyShane) [reply src/features/broker-protection/actions/fill-form.js:45]: @muodov Yep I agree - due the isolated nature of the rest of this code, I'd prefer to come back and move this after the merge, so you think that's ok?

File: src/features/broker-protection/actions/extract.js
Comment (shakyShane) [reply src/features/broker-protection/actions/extract.js:80]: I altered the types + added a comment to make it explicit. 

IMO, this needs re-visiting at a later date, there's quite a mix of types on inputs/output and it's making it difficult to review.

File: package.json
Comment (shakyShane) [reply package.json]: done.

**Comment by shakyShane** [comment]
@jonathanKingston @muodov - I've added a few extra commits to address the feedback, thanks!

I think we'll ready to merge this, and then it's onto the integration tests :)

**Comment by brianhall** [comment]
@shakyShane (Adding this from mobile, so apologies if I’m not seeing this change and for the late comment). We spoke on Asana about adding playwright as a top level dependency, can we do that in this PR?

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (muodov) [inline src/features/broker-protection/actions/build-url-transforms.js:139]: do we need to decode `encodedValue`?

**Review by muodov — APPROVED** [review]
looks good, just one remark about url encoding. Thanks for the changes!

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url-transforms.js:139]: Ahh sorry I missed this!

We don't need to, anything that gets this far has already been through the URL constructor, which decodes.

Comment (muodov) [reply src/features/broker-protection/actions/build-url-transforms.js:139]: wait, why do we need the `%7B` and `%7D` bits in the regex then?

**Comment by shakyShane** [comment]
Thanks! merging now since we have approvals from @brianhall and @muodov, I also addressed feedback from @jonathanKingston

File: src/features/broker-protection/actions/build-url-transforms.js
Comment (shakyShane) [reply src/features/broker-protection/actions/build-url-transforms.js:139]: This is because pathname and query segments are treated differently when passed through the URL constructor, so this handles both situations.

if this causes confusion, we could split it up, and make it clear which we're processing.

**Pull request merged**
